crates/external_services/src/grpc_client/dynamic_routing.rs

```diff
@@ -0,0 +1,265 @@
+use std::fmt::Debug;
+
+use api_models::routing::{
+    CurrentBlockThreshold, RoutableConnectorChoice, RoutableConnectorChoiceWithStatus,
+    SuccessBasedRoutingConfig, SuccessBasedRoutingConfigBody,
+};
+use common_utils::{errors::CustomResult, ext_traits::OptionExt, transformers::ForeignTryFrom};
+use error_stack::ResultExt;
+use serde;
+use success_rate::{
+    success_rate_calculator_client::SuccessRateCalculatorClient, CalSuccessRateConfig,
+    CalSuccessRateRequest, CalSuccessRateResponse,
+    CurrentBlockThreshold as DynamicCurrentThreshold, LabelWithStatus,
+    UpdateSuccessRateWindowConfig, UpdateSuccessRateWindowRequest, UpdateSuccessRateWindowResponse,
+};
+use tonic::transport::Channel;
+#[allow(
+    missing_docs,
+    unused_qualifications,
+    clippy::unwrap_used,
+    clippy::as_conversions
+)]
+pub mod success_rate {
+    tonic::include_proto!("success_rate");
+}
+/// Result type for Dynamic Routing
+pub type DynamicRoutingResult<T> = CustomResult<T, DynamicRoutingError>;
+
+/// Dynamic Routing Errors
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum DynamicRoutingError {
+    /// The required input is missing
+    #[error("Missing Required Field : {field} for building the Dynamic Routing Request")]
+    MissingRequiredField {
+        /// The required field name
+        field: String,
+    },
+    /// Error from Dynamic Routing Server
+    #[error("Error from Dynamic Routing Server : {0}")]
+    SuccessRateBasedRoutingFailure(String),
+}
+
+/// Type that consists of all the services provided by the client
+#[derive(Debug, Clone)]
+pub struct RoutingStrategy {
+    /// success rate service for Dynamic Routing
+    pub success_rate_client: Option<SuccessRateCalculatorClient<Channel>>,
+}
+
+/// Contains the Dynamic Routing Client Config
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize, Default)]
+#[serde(untagged)]
+pub enum DynamicRoutingClientConfig {
+    /// If the dynamic routing client config has been enabled
+    Enabled {
+        /// The host for the client
+        host: String,
+        /// The port of the client
+        port: u16,
+    },
+    #[default]
+    /// If the dynamic routing client config has been disabled
+    Disabled,
+}
+
+impl DynamicRoutingClientConfig {
+    /// establish connection with the server
+    pub async fn get_dynamic_routing_connection(
+        self,
+    ) -> Result<RoutingStrategy, Box<dyn std::error::Error>> {
+        let success_rate_client = match self {
+            Self::Enabled { host, port } => {
+                let uri = format!("http://{}:{}", host, port);
```

Comment:
Do we ever plan to have TLS support? Asking since the scheme is hardcoded as `http`.